server/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java

```diff
@@ -350,22 +351,25 @@ public void testIndexWithUnknownSetting() throws Exception {
         }
 
         final IndexMetaData metaData = state.getMetaData().index("test");
-        internalCluster().fullRestart(new InternalTestCluster.RestartCallback() {
+        internalCluster().fullRestart(new InternalTestCluster.FullRestartCallback() {
+
             @Override
-            public Settings onNodeStopped(String nodeName) throws Exception {
-                if (dataOrMasterNodeNames.contains(nodeName)) {
-                    final NodeEnvironment nodeEnvironment = internalCluster().getInstance(NodeEnvironment.class, nodeName);
-                    final IndexMetaData brokenMetaData =
-                            IndexMetaData
-                                    .builder(metaData)
-                                    .settings(Settings.builder().put(metaData.getSettings()).put("index.foo", true))
-                                    .build();
-                    // so evil
-                    IndexMetaData.FORMAT.write(brokenMetaData, nodeEnvironment.indexPaths(brokenMetaData.getIndex()));
+            public void onAllNodesStopped(List<String> nodeNames) throws Exception {
```

Comment:
Because the test was failing sometimes if onNodeStopped callback was used. The reason is in a way fullRestart works.

1. It stops all nodes and re-shuffles data paths between them.
2. It calls onNodeStopped callback for the node and then re-initializes the node. This is done for each node one by one.
3. It starts all the nodes.

The problem is in 2. Since data paths are re-shuffled, we can concurrently write broken metadata inside the test and write metadata on the startup (inside GatewatMetaState constructor). So the safest thing to do is to break metadata when all nodes are stopped.

Reply:
I think the proper fix here would be for `fullRestart` to call `onNodeStopped` on all the nodes when they are shut down, not when the cluster is starting back up. There is no need to introduce another abstraction here.

Reply:
It makes sense. I've done it in 3 commits:

1. Revert FullRestartCallback commit https://github.com/elastic/elasticsearch/pull/35049/commits/ffd25a64e475073598bb3f3c035746378dd88d60
2. Full restart should call onNodeStopped before re-creating the nodes https://github.com/elastic/elasticsearch/pull/35049/commits/f21ddf9538bc0750f3ec6911039c058013243412
3. Fix broken metadata tests https://github.com/elastic/elasticsearch/pull/35049/commits/f1ff7325f3b809d678e8cfa58641aec8f6c73eba

Reply:
I've opened https://github.com/elastic/elasticsearch/pull/35494 against master which does a similar kind of refactoring of `onNodeStopped`, but also goes a little further with the refactoring to make better reuse of the existing methods while also enforcing that a (temporarily) closed node (+ client) is not being used.

Reply:
Ok, I'll review it tomorrow and once it's merged to master and back to my branch, I will revert the corresponding commit.

Reply:
I've reverted my commit https://github.com/elastic/elasticsearch/pull/35049/commits/17bb22a0f677f5680ffe688d35fe7b055b4c2c82 and merged master into zen2 and zen2 into my branch https://github.com/elastic/elasticsearch/pull/35049/commits/a5de8ad8365a0a9b201103305ba3152587a3e853